proposals/dnd-list.md

```diff
@@ -0,0 +1,133 @@
+
+export default () =>
+  <StatefulList
+    initialState={{items: ['Item 1', 'Item 2', 'Item 3']}}
+    onChange={({newState, oldIndex, newIndex}) => {}}
+  />
+```
+
+### Advanced usage
+
+```js
+import * as React from 'react';
+import {
+  StatefulList,
+  StyledList,
+  StyledItem
+} from 'baseui/dnd-list';
+import {styled} from 'baseui';
+
+CustomList = styled(StyledList, props => ({
+  border: '1px solid green'
+}));
+
+CustomItem = styled(StyledItem, props => ({
+  borderBottom: props.$dragged ? `1px solid red` : 'none'
+}));
+
+export default () =>
+  <StatefulList
+    initialState={{ items: ['Item 1', 'Item 2', 'Item 3']}}
+    overrides={{
+      List: CustomList,
+      Item: CustomItem
+    }}
+  >
+```
+
+## Exports
+
+* `StatefulList`
+* `List`
+* `StyledList`
+* `StyledItem`
+* `StyledDragHandle`
+* `StyledCloseHandle`
+* `STATE_CHANGE_TYPE`
+
+## `List` API
+
+* `items: Array<string>` - Required.
+  List items.
+* `overrides: {Root, Item, DragHandle, CloseHandle}` - Optional.
+  Overrides for presentational components. See "Presentational Components Props API" below
+  * `[ComponentName]: ReactComponent | {props: {}, style: {}, component: ReactComponent}` - Optional
+* `onChange: ({oldIndex: number, newIndex: number})` - Optional.
+  onChange handler is called every time an item is moved or removed (newIndex = -1)
+
+## `StatefulList` API
+
+* `initialState: {items: Array<string>}` - Required.
+  Initial state of the component where `items` are list items.
+* `overrides: {Root, List, Item, DragHandle, CloseHandle}` - Optional.
+  Overrides for presentational components. See "Presentational Components Props API" below
+  * `[ComponentName]: ReactComponent | {props: {}, style: {}, component: ReactComponent}` - Optional
+* `onChange: ({newState: Array<string>, oldIndex: number, newIndex: number})` - Optional.
+  onChange handler that is called every time an item is moved or removed (newIndex = -1).
+* `stateReducer: (type: string, oldIndex: number, newIndex: number, currentState: {items: Array<string>}) => stateToSet: {items: Array<string>}` - Optional. Useful when you want to replace default array-move logic.
+  A state change handler.
+  * `type` - a state change type
+  * `oldIndex` - an original item index
+  * `newIndex` - a new item index (after being moved or removed when -1)
+  * `currentState` - current state value
+  * `stateToSet` - a return value that the state will be updated with
+
+## Presentational components props API
+
+These properties are passed to presentational (styled) components that are exported:
+
+* `$dragged: boolean` If a particular item is active
+* `$hovered: boolean` If a particular item is hovered
+* `$focused: boolean` If a particular item is focused (`tab`, `shift + tab`)
+
+## STATE_CHANGE_TYPE Constant
+
+* `change` - An item was moved or removed
+
+## Accessibility
+
+`tab`, `shift + tab` - to move focus between tabs.
+
+`Space` - to drag (focus) or drop an item
```

Comment:
according to medium article `Esc` should cancel the move.

Somewhat unclear in this document is what steps one would take to change order with keyboard. Is it something like below?

```
tab to item
space to begin item drag
arrow up/down to change position
space to end item drag
```